File: src/main/java/me/viniciusarnhold/altaria/command/pool/PoolManager.java
package me.viniciusarnhold.altaria.command.pool;

import com.vdurmont.emoji.EmojiManager;
import me.viniciusarnhold.altaria.command.MessageUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.handle.obj.IMessage;
import sx.blah.discord.handle.obj.IReaction;
import sx.blah.discord.handle.obj.IUser;
import sx.blah.discord.util.DiscordException;
import sx.blah.discord.util.EmbedBuilder;
import sx.blah.discord.util.MissingPermissionsException;
import sx.blah.discord.util.RateLimitException;

import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static me.viniciusarnhold.altaria.utils.Timers.messageDeletionService;

/**
 * Created by Vinicius.
 *
 * @since ${PROJECT_VERSION}
 */
public class PoolManager {

    private static final Logger logger = LogManager.getLogger();

    private static final PoolManager ourInstance = new PoolManager();

    private PoolManager() {

    }

    @NotNull
    public static PoolManager getInstance() {
        return ourInstance;
    }

    public final void registerPool(@NotNull IMessage message, @NotNull Pool pool) {
        logger.trace("Registering new pool {}, called by {} for {} seconds.",
                () -> pool.id(),
                () -> message.getAuthor().getName(),
                () -> TimeUnit.MILLISECONDS.convert(pool.time(), TimeUnit.SECONDS));

        new Timer("Pool timer" + pool.id(), true)
                .schedule(new PoolFinalizerTask(pool), TimeUnit.MILLISECONDS.convert(pool.time(), TimeUnit.SECONDS));
    }


    private static class PoolFinalizerTask extends TimerTask {

        private static final Logger logger = LogManager.getLogger();

        private final Pool pool;

        public PoolFinalizerTask(Pool pool) {
            this.pool = pool;
        }

        /**
         * The action to be performed by this timer task.
         */
        @Override
        public void run() {
            IChannel channel = this.pool.message().getChannel();
            IMessage currentMessage = channel.getMessageByID(this.pool.message().getID());
            if (currentMessage == null) {
                logger.trace("Pool {} finished but the message was Deleted", pool.id());
                return;
            }

            final Map<String, String> options = pool.options();
            final Map<String, List<IUser>> results = new HashMap<>();


            if (pool.type() == Pool.Type.MULTI) {
                for (Map.Entry<String, String> entry : options.entrySet()) {
                    IReaction reaction = currentMessage.getReactionByName(EmojiManager.getForAlias(entry.getKey()).getUnicode());
                    if (reaction != null) {
                        try {
                            results.put(entry.getKey(), reaction.getUsers().stream().filter(u -> !u.isBot()).collect(Collectors.toList()));
                        } catch (@NotNull RateLimitException | DiscordException e) {
                            logger.error(e);
                            return;
                        }
                    }
                }
            } else {
                Set<IUser> users = new HashSet<>();
                for (Map.Entry<String, String> entry : options.entrySet()) {
                    IReaction reaction = currentMessage.getReactionByName(EmojiManager.getForAlias(entry.getKey()).getUnicode());
                    if (reaction != null) {
                        List<IUser> rUsers = Collections.emptyList();
                        try {
                            rUsers = reaction.getUsers().stream().filter(u -> !u.isBot()).collect(Collectors.toList());
                        } catch (@NotNull RateLimitException | DiscordException e) {
                            logger.error(e);
                        }
                        rUsers.removeAll(users);
                        results.put(entry.getKey(), rUsers);
                        users.addAll(rUsers);
                    }
                }
            }
            final EmbedBuilder builder = MessageUtils.getEmbedBuilder(currentMessage.getAuthor())
                    .withTitle("Pool Results");

            results.entrySet().stream()
                    .sorted((o1, o2) -> o2.getValue().size() - o1.getValue().size())
                    .filter((e) -> e.getValue().size() > 0)
                    .forEachOrdered(e ->
                            builder.appendField(
                                    e.getKey() + " - " + options.get(e.getKey()),
                                    e.getValue().stream()
                                            .map(IUser::getName)
                                            .limit(15) //TODO Check if 15 is enough or if a embed message can fit more than 15 icons per line
                                            .collect(Collectors.joining(", ")),
                                    false));

            builder.withDescription("Winner: " + results.entrySet().stream()
                    .max(Comparator.comparingInt(e -> e.getValue().size()))
                    .map(e -> e.getKey() + " - " + options.get(e.getKey()))
                    .orElse("No Votes"));

            try {
                messageDeletionService().schedule(
                        MessageUtils.getMessageBuilder(currentMessage.getChannel())
                                .withEmbed(builder.build())
                                .send());

                pool.message().delete();

            } catch (@NotNull RateLimitException | MissingPermissionsException | DiscordException e) {
                MessageUtils.handleDiscord4JException(logger, e, PoolCommand.getInstance(), currentMessage);
            }
        }
    }
}

File: commands/src/main/kotlin/me/viniciusarnhold/altaria/commands/pool/PoolManager.kt
package me.viniciusarnhold.altaria.commands.pool

import com.vdurmont.emoji.EmojiManager
import me.viniciusarnhold.altaria.api.MessageUtils
import me.viniciusarnhold.altaria.utils.Timers
import org.apache.logging.log4j.LogManager
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.handle.obj.IUser
import sx.blah.discord.util.DiscordException
import sx.blah.discord.util.MissingPermissionsException
import sx.blah.discord.util.RateLimitException
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by Vinicius.

 * @since 1.0
 */
class PoolManager private constructor() {

    fun registerPool(command: PoolCommand, message: IMessage, pool: Pool) {
        logger.trace("Registering new pool {}, called by {} for {} seconds.",
                { pool.id() },
                { message.author.name },
                { TimeUnit.MILLISECONDS.convert(pool.time(), TimeUnit.SECONDS) })

        Timer("Pool timer" + pool.id(), true)
                .schedule(PoolFinalizerTask(command, pool), TimeUnit.MILLISECONDS.convert(pool.time(), TimeUnit.SECONDS))
    }


    private class PoolFinalizerTask(private val command: PoolCommand, private val pool: Pool) : TimerTask() {

        /**
         * The action to be performed by this timer task.
         */
        override fun run() {
            val channel = this.pool.message().channel
            val currentMessage = channel.getMessageByID(this.pool.message().longID)
            if (currentMessage == null) {
                logger.trace("Pool {} finished but the message was Deleted", pool.id())
                return
            }

            val options = pool.options()
            val results = HashMap<String, List<IUser>>()


            if (pool.type() == Pool.Type.MULTI) {
                for ((key) in options) {
                    val reaction = currentMessage.getReactionByUnicode(EmojiManager.getForAlias(key).unicode)
                    if (reaction != null) {
                        try {
                            results.put(key, reaction.users.filter { !it.isBot })
                        } catch (e: RateLimitException) {
                            logger.error(e)
                            return
                        } catch (e: DiscordException) {
                            logger.error(e)
                            return
                        }

                    }
                }
            } else {
                val users = HashSet<IUser>()
                for ((key) in options) {
                    val reaction = currentMessage.getReactionByUnicode(EmojiManager.getForAlias(key).unicode)
                    if (reaction != null) {
                        val rUsers = try {
                            reaction.users.filter { !it.isBot }.toMutableList()
                        } catch (e: RateLimitException) {
                            logger.error(e)
                            mutableListOf<IUser>()
                        } catch (e: DiscordException) {
                            logger.error(e)
                            mutableListOf<IUser>()
                        }

                        rUsers.removeAll(users)
                        results.put(key, rUsers)
                        users.addAll(rUsers)
                    }
                }
            }
            val builder = MessageUtils.getEmbedBuilder(currentMessage.author)
                    .withTitle("Pool Results")

            results.entries.stream()
                    .sorted { o1, o2 -> o2.value.size - o1.value.size }
                    .filter { e -> e.value.size > 0 }
                    .forEachOrdered { e ->
                        builder.appendField(
                                e.key + " - " + options[e.key],
                                e.value.asSequence()
                                        .map(transform = { it.name })
                                        .take(n = 15)
                                        .joinToString(separator = ", "),
                                false)
                    }

            val winner = results.entries.maxWith(comparator = kotlin.Comparator.comparingInt { e -> e.value.size })?.run({
                "$key - ${options[key]}"
            }) ?: "No Votes"

            builder.withDescription("Winner: $winner")

            try {
                Timers.MessageDeletionService.schedule(
                        MessageUtils.getMessageBuilder(currentMessage.channel)
                                .withEmbed(builder.build())
                                .send())

                pool.message().delete()

            } catch (e: RateLimitException) {
                MessageUtils.handleDiscord4JException(logger, e, command, currentMessage)
            } catch (e: MissingPermissionsException) {
                MessageUtils.handleDiscord4JException(logger, e, command, currentMessage)
            } catch (e: DiscordException) {
                MessageUtils.handleDiscord4JException(logger, e, command, currentMessage)
            }

        }

        companion object {

            private val logger = LogManager.getLogger()
        }
    }

    companion object {

        private val logger = LogManager.getLogger()

        val instance = PoolManager()
    }
}
